tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_collision.c

```diff
@@ -0,0 +1,230 @@
+/* main.c - Application main entry point */
+
+/*
+ * Copyright (c) 2022 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stddef.h>
+
+#include <zephyr/types.h>
+#include <sys/printk.h>
+#include <sys/util.h>
+#include <sys/byteorder.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/att.h>
+
+#include "bs_types.h"
+#include "bs_tracing.h"
+#include "bstests.h"
+
+extern enum bst_result_t bst_result;
+
+#define FAIL(...)                                                                                  \
+	do {                                                                                       \
+		bst_result = Failed;                                                               \
+		bs_trace_error_time_line(__VA_ARGS__);                                             \
+	} while (0)
+
+#define PASS(...)                                                                                  \
+	do {                                                                                       \
+		bst_result = Passed;                                                               \
+		bs_trace_info_time(1, __VA_ARGS__);                                                \
+	} while (0)
+
+static struct bt_conn *default_conn;
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+};
+
+static bool volatile is_connected;
+
+static void connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+		FAIL("Failed to connect to %s (%u)\n", addr, conn_err);
+	}
+
+	default_conn = conn;
+	printk("Connected: %s\n", addr);
+	is_connected = true;
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	struct bt_conn_info info;
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+	(void)bt_conn_get_info(conn, &info);
+
+	printk("Disconnected: %s (reason 0x%02x)\n", addr, reason);
+
+	if (default_conn != conn) {
+		FAIL("Conn mismatch disconnect %s %s)\n", default_conn, conn);
+	}
+
+	if (info.role == BT_CONN_ROLE_CENTRAL) {
+		bt_conn_unref(default_conn);
+	}
+
+	default_conn = NULL;
```

Comment:
Once you've fixed the `default_conn` reference counting there should be an unconditional `bt_conn_unref(default_conn)` before setting it to NULL (and also remove the conditional unref on line 78)